### PR TITLE
Support for PubSub Messaging

### PR DESCRIPTION
**Pull request by purinda**
# Pub/Sub

Pub/Sub mechanism implements the message passing between applications using the ORM leveraging the native implementations of Pub/Sub within the database driver in use.

An example use case for Pub/Sub would be sending messages to background tasks that run on a separate process apart from the application itself, processing the workload without holding the application process.

### CODE REVIEW

File: src/Bravo3/Orm/Drivers/Redis/RedisDriver.php [hunk 1]
```diff
@@ -91,6 +92,10 @@ public function __construct(
             $this->client = $client;
         }
 
+        // Enable Pub/Sub mechanism
+        if ($this->isPubSubSupported()) {
```
Comment (trq) [inline src/Bravo3/Orm/Drivers/Redis/RedisDriver.php]: Should this be here?

File: docs/PubSub.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+Following examples use a simple string as the message, in practice it is possible to use JSON encoded strings or serialised data structures to be passed via the PubSub mechanism (if the underlying driver permits).
+
+### Subscriber example
+
+``` sourceCode
+$channel_name = 'bg-task-1';
+
+// Send the message to the subscriber
+$pubsub_manager->addListener($channel_name, function(PubSubEvent $e) {
+    list($cmd, $msg) = explode(' ', $e->getMessage());
```
Comment (y2khjh) [inline docs/PubSub.md:45]: So can't use space in message?

File: src/Bravo3/Orm/Drivers/Redis/RedisDriver.php (hunk 1 above)
Comment (purinda) [inline src/Bravo3/Orm/Drivers/Redis/RedisDriver.php]: No, thanks, removed.

File: docs/PubSub.md (hunk 2 above)
Comment (purinda) [inline docs/PubSub.md:45]: Just an example @y2khjh, in that example you can't.

File: src/Bravo3/Orm/Drivers/PubSubDriverInterface.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    public function listenToPubSub(callable $callback);
+
+    /**
+     * Publishes a message to the configured channel. Channel and Message length is limited based on the driver used.
+     *
+     * @param string $channel
+     * @param string $message
+     * @return int
+     */
+    public function publishMessage($channel, $message);
```
Comment (y2khjh) [inline src/Bravo3/Orm/Drivers/PubSubDriverInterface.php:46]: Unittest?

File: src/Bravo3/Orm/Drivers/Redis/RedisDriver.php [hunk 4]
```diff
@@ -686,4 +687,57 @@ public function calculateRetryDelay($retry_iteration)
             return 0;
         }
     }
+
+    /**
+     * Checks if the driver supports publish/subscribe pattern.
+     *
+     * @return bool
+     */
+    public function isPubSubSupported()
```
Comment (trq) [inline src/Bravo3/Orm/Drivers/Redis/RedisDriver.php:695]: Needs to be part of the PubSubDriverInterface interface

File: src/Bravo3/Orm/Drivers/PubSubDriverInterface.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+interface PubSubDriverInterface
+{
+    const SUBSCRIPTION_PATTERN = 'BRAVO3_EVT-*';
+
+    /**
+     * Start listening to subscribed channels of the database (if PubSub is supported)
+     * @return void
+     */
+    public function listenToPubSub(callable $callback);
```
Comment (y2khjh) [inline src/Bravo3/Orm/Drivers/PubSubDriverInterface.php:37]: Unittest?

File: src/Bravo3/Orm/Drivers/Redis/RedisDriver.php (hunk 4 above)
Comment (purinda) [inline src/Bravo3/Orm/Drivers/Redis/RedisDriver.php:695]: Added. Thanks

File: src/Bravo3/Orm/Drivers/PubSubDriverInterface.php (hunk 3 above)
Comment (purinda) [inline src/Bravo3/Orm/Drivers/PubSubDriverInterface.php:46]: Explain a bit more? Unit tests should be for the driver, which we assume works as it is from Predis.

**Pull request merged**